elastic/kibana
# [Uptime] Redirect to error page when Heartbeat mappings are missing

## Summary

Resolves https://github.com/elastic/uptime/issues/359.

This is ~WIP~ code resulting from a spike. It requires design input and further refinement from a code standpoint, as well as some copy updates.

This change would add a new page, mapping error, to the Uptime app's routes list. It should include helpful information that lets the user know that they need to fix their indices when Heartbeat's mappings were not appropriately installed.

At the moment, I've only modified one component on the Overview page to perform the redirect. Essentially, we catch this specific error from Elasticsearch and return it to the client; all other errors remain in a thrown state.

![image](https://user-images.githubusercontent.com/18429259/131718889-d60820a2-b325-4f58-a828-3b2332b2e9d7.png)

## Testing

1. Start Elasticsearch and Heartbeat
1. `curl -u elastic:{PASSWORD} -X DELETE "http://localhost:9200/{HEARTBEAT_INDEX}"`
1. Navigate to the Uptime UI, you should be redirected to the mapping error page
1. Kill Heartbeat
1. `curl -u elastic:{PASSWORD} -X DELETE "http://localhost:9200/{HEARTBEAT_INDEX}"`
1. Start Heartbeat
1. Navigate to Uptime UI, you should see the overview page per usual

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/master/RISK_MATRIX.mdx) |


### For maintainers
